Fix list formatting in kubeadm setup tutorial

tfogo opened the pull request:
There is an issue with ordered lists in the _Creating HA clusters with
kubeadm_ setup tutorial. All list items are labeled 1. This is because
the underlying markdown processor for Hugo, blackfriday, requires
list items with multiple paragraphs to have four space indentation.
(See russross/blackfriday#244)

This commit adds an extra space before paragraphs in lists so they are
correctly interpreted as multi-paragraph lists.

However, this exposes a bug in blackfriday where lines beginning with
`-` in codefenced code blocks inside lists are parsed as sub-lists.
This means code fenced code blocks inside lists that contain yaml are
malformed. (See russross/blackfriday#239)

So this commit mitigates this bug in two situations. It uses a 4-space
indented code block in one situation. In another situation where a list
only contained one element, it is no longer a list.

tfogo commented:
/assign @mistyhacks

neolit123 commented:
yikes, this is horrible. not your efforts @tfogo, but rather the underlying bugs...
how about not using ordered lists at all if they are that "demanding" / buggy and use unordered lists or manual ordering (somehow) until the issues are resolved at the markdown processor?

@tfogo @mistyhacks WDYT?

tfogo commented:
@neolit123 I know! It's a pretty awkward bug, particularly for our yaml heavy use case. There's been open PR to fix it for several months (https://github.com/russross/blackfriday/pull/372).

Unordered lists are also affected by the bug, unfortunately. 

I think manual ordering might work well in some cases. We could manually number each item in the list, but multi-paragraph list items won't be properly indented. Which means they aren't visually grouped together. Alternatively, we could signpost with bolded text (e.g. **Step 1:**, **Step 2:**).

Whatever is decided, it might be useful to add a note to the style guide (similar to the _Common Shortcode Issues_ section).

neolit123 commented:
> Alternatively, we could signpost with bolded text (e.g. Step 1:, Step 2:).

i personally i like this solution. if the ordering is broken at the backend we should write the steps ourselves.
more difficult to maintain, yet the steps aren't that many.

mistyhacks commented:
There is a slightly cleaner way to fix this. Let me add a commit.

mistyhacks commented:
OK, I added my commit. What I did was this:

- Always line up text after the step numbers with tabstops. This helps with debugging later, even though it means that single-digit steps numbers get an extra space after the period.
- Indent all block-level things within list items to an extra tabstop. In my testing for the new smoke test page, this was the way to get the most consistent results.
- While I was in there, I wrapped some admonitions in the correct short tags, rewrapped some long lines, and did light rewording in places that really stood out to me.
- I also fixed some of the code fences so that they don't break Atom so badly when editing the files there. This has to do with wrapping things in shell that don't look like shell output (like all the HEREDOC stuff, especially HEREDOC that is YAML or JSON). This is purely to help future editors of this file.

mistyhacks commented:
/lgtm

@tfogo please ack my content changes here. The output looks fine to me now, though the sub-lists look kind of ugly. If you agree, maybe file a GH issue for CSS.

tfogo commented:
@mistyhacks Thanks, looks cleaner for sure!

One awkward thing about indenting block-level things inside lists an extra tabstop is that code blocks all have two spaces before each line. This is a bit awkward for copy-paste. Particularly with commands that are piping a heredoc to a file. Do you have any thoughts on this?

tfogo commented on content/en/docs/setup/independent/high-availability.md:
```diff
@@ -470,40 +491,48 @@ Only follow this step if your etcd is hosted on dedicated nodes (**Option 1**).
 
-    **Note:** If you are using Kubernetes 1.9+, you can replace the `apiserver-count: 3` extra argument with `endpoint-reconciler-type: lease`. For more information, see [the documentation](/docs/admin/high-availability/#endpoint-reconciler).
+      {{< note >}}
+      **Note:** If you are using Kubernetes 1.9+, you can replace the `apiserver-count: 3` extra argument with `endpoint-reconciler-type: lease`. For more information, see [the documentation](/docs/admin/high-availability/#endpoint-reconciler).
```
Same issue with this one.

tfogo commented:
Added a fix for the shortcodes - I know it's ugly that they're on the same line. If you know of a better fix, let me know :)

neolit123 commented on content/en/docs/setup/independent/high-availability.md:
````diff
@@ -53,296 +53,304 @@ For **Option 2**: you can skip to the next step. Any reference to `etcd0`, `etcd
+      ```shell
+      export PEER_NAME=$(hostname)
+      export PRIVATE_IP=$(ip addr show eth1 | grep -Po 'inet \K[\d.]+')
+      ```
````
can we just leave everything aligned to the left? i don't think it looks bad without it.
once the ordered list bugs are fixed we just have to modify the digits and the extra space and the alignment will be auto-added; if i understand this correctly, that is...

tfogo replied:
I don't think leaving everything aligned to the left is necessary. The sublist bug only affects two code blocks and we've edited it in such a way that they're no longer affected. So we have the list indentation being rendered correctly and there are no sublist bugs. 

And we have no idea when a fix to this bug will make its way into Hugo.

mistyhacks replied:
Leaving everything aligned to the left seems to sometimes work and sometimes not. Indenting everything a level seems to always work. I think we have to teach contributors how to contribute docs in the way that creates the least amount of cognitive load, so it's easier to give guidance to always indent things inside lists another level.

mistyhacks commented on content/en/docs/setup/independent/high-availability.md:
```diff
@@ -53,296 +53,304 @@ For **Option 2**: you can skip to the next step. Any reference to `etcd0`, `etcd
 {{< /tabs >}}
 
 
+
```
Nit: This probably can be reverted (did I do this?)

tfogo replied:
My bad! I'll fix that.

tfogo commented:
@mistyhacks I've spoken with the author of this PR (https://github.com/russross/blackfriday/pull/372) to fix the lists inside code blocks issue with blackfriday. I'm hopeful that we can get that fixed merged in soon.

tfogo commented:
@mistyhacks I fixed that newline and rebased.

Do you have any thoughts on the code blocks having an extra two spaces? I mentioned above that it can be awkward for copy-pasting heredocs. I don't think it's a huge issue - but it might be good to revisit it when blackfriday is acting more predictable.

zacharysarah commented:
More on removing code formatting from headers: this is what was happening.
![screen shot 2018-06-04 at 12 43 47 pm](https://user-images.githubusercontent.com/3210446/40938053-7ff1f6b4-67f5-11e8-9ccc-9e414c7f9427.png)

@mistyhacks Any recommendation about how to proceed here?

mistyhacks commented:
I thought @bep fixed this in #8797? I don't think you should have to remove them.

Bradamant3 commented:
@tfogo if you could please rebase/resolve conflicts, we can get this merged.  It also looks as though your HTML comment about code fencing appears in the config file example -- please remove :-)

mistyhacks commented:
This became a mess because a lot of these fixes have actually already been made in master. I worked with @Bradamant3 to figure out which changes were unique to this PR and I re-applied those, and also fixed an issue I saw with the tabs in the content and another with a skip in heading order level.

Bradamant3 commented:
/lgtm
/approve

tfogo commented:
@mistyhacks @Bradamant3 Thank you for going through and doing that. 

FYI, we've fixed the blackfriday issue in v2 (russross/blackfriday#372) but still need to backport to v1 for Hugo.
